02-household-queries/ingest.py

```diff
@@ -0,0 +1,111 @@
+from langchain_google_genai import ChatGoogleGenerativeAI, GoogleGenerativeAIEmbeddings
+from langchain.prompts import PromptTemplate
+from langchain.llms import GPT4All
+from langchain import LLMChain
+from bs4 import BeautifulSoup
+from langchain_community.document_loaders import JSONLoader, PyPDFLoader
+from langchain.docstore.document import Document
+from langchain_text_splitters import RecursiveCharacterTextSplitter
+from langchain.chains.combine_documents import create_stuff_documents_chain
+from langchain.chains import create_retrieval_chain
+from langchain.vectorstores import Chroma
+from langchain.embeddings.sentence_transformer import SentenceTransformerEmbeddings
+import json
+import os
+from pathlib import Path
+from pprint import pprint
+import dotenv
+
+# Implementation with Google Gemini embeddings and storing in Chroma
+# dotenv.load_dotenv()
+# #Load the models
+# GOOGLE_API_KEY = os.environ.get('GOOGLE_API_KEY')
+
+# llm = ChatGoogleGenerativeAI(model="gemini-pro",
+#                              verbose = True,temperature = 0.1,google_api_key=GOOGLE_API_KEY)
+# embeddings = GoogleGenerativeAIEmbeddings(model="models/embedding-001")
+
+# Load the PDF and create chunks
+# download from https://nava.slack.com/archives/C06ETE82UHM/p1710880671273909?thread_ts=1710880610.675809&cid=C06ETE82UHM
+# loader = PyPDFLoader("./7_cfr_part_272__up_to_date_as_of_3-15-2024_.pdf")
+# vectordb=Chroma.from_documents(pdf_pages,embeddings)
+
+# pdf_pages = loader.load_and_split(text_splitter)
+
+#Configure Chroma as a retriever with top_k=5
+# retriever = vectordb.as_retriever(search_kwargs={"k": 5})
+
+#Create the retrieval chain
+# template = """
+# You are a helpful AI assistant.
+# Answer based on the context provided. 
+# context: {context}
+# input: {input}
+# answer:
+# """
+# prompt = PromptTemplate.from_template(template)
+# combine_docs_chain = create_stuff_documents_chain(llm, prompt)
+# retrieval_chain = create_retrieval_chain(retriever, combine_docs_chain)
+
+# Invoke the retrieval chain
+# response=retrieval_chain.invoke({"input":"what are appeals?"})
+
+# Print the answer to the question
+# print(response["answer"])
+
+
+# Load the json and create chunks
+# download from https://drive.google.com/file/d/1UoWmktXS5nqgIWj2x_O5hgzwU0yVuaJc/view
+guru_file_path='./guru_cards_for_nava.json'
+
+
+def get_text_chunks_langchain(text):
+    text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=100)
+    texts = text_splitter.split_text(text)
+    docs = [Document(page_content=t) for t in texts]
+    return docs
+
+embeddings = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")
```

Comment:
Should check and notify when a sentence gets truncated.
> "all-MiniLM-L6-v2" encodes texts [up to 256 words](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2#:~:text=By%20default%2C%20input%20text%20longer%20than%20256%20word%20pieces%20is%20truncated.). It’ll truncate any text longer than this.

https://realpython.com/chromadb-vector-database/

Reply:
The `RecursiveCharacterTextSplitter` can limit the chunk size, but character length (the measure for chunk size) doesn't cleanly map to word length.